gap/weights.gi

```diff
@@ -166,3 +166,453 @@ function(digraph)
+    outNeighbours := mutableOuts[u];
+    for idx in [1 .. Size(outNeighbours)] do
+      v                      := outNeighbours[idx];
+      w                      := mutableWeights[u][idx];
+      mutableWeights[u][idx] := w +
+                                bellmanDistances[u] - bellmanDistances[v];
+    od;
+  od;
+
+  Remove(mutableOuts, new);
+  Remove(mutableWeights, new);
+
+  digraph   := EdgeWeightedDigraph(mutableOuts, mutableWeights);
+  distances := EmptyPlist(nrVertices);
+  parents   := EmptyPlist(nrVertices);
+  edges     := EmptyPlist(nrVertices);
+
+  # run dijkstra
+  for u in vertices do
+    dijkstra     := DIGRAPHS_Edge_Weighted_Dijkstra(digraph, u);
+    distances[u] := dijkstra.distances;
+    parents[u]   := dijkstra.parents;
+    edges[u]     := dijkstra.edges;
+  od;
+
+  # correct distances
+  for u in vertices do
+    for v in vertices do
+      if distances[u][v] = fail then
+        continue;
+      fi;
+      distances[u][v] := distances[u][v] +
+                         bellmanDistances[v] - bellmanDistances[u];
+    od;
+  od;
+
+  return rec(distances := distances, parents := parents, edges := edges);
+end);
+
+InstallGlobalFunction(DIGRAPHS_Edge_Weighted_FloydWarshall,
+function(digraph)
+  local weights, adjMatrix, vertices, nrVertices, u, v, edges, outs, idx,
+        outNeighbours, w, i, k, distances, parents;
+  weights    := EdgeWeights(digraph);
+  vertices   := DigraphVertices(digraph);
+  nrVertices := Size(vertices);
+  outs       := OutNeighbours(digraph);
+
+  # Create adjacency matrix with (minimum weight, edge index), or a hole
+  adjMatrix := EmptyPlist(nrVertices);
+  for u in vertices do
+    adjMatrix[u] := EmptyPlist(nrVertices);
+    outNeighbours := outs[u];
+    for idx in [1 .. Size(outNeighbours)] do
+      v := outNeighbours[idx];  # the out neighbour
+      w := weights[u][idx];     # the weight to the out neighbour
+      # Use minimum weight edge
+      if (not IsBound(adjMatrix[u][v])) or (w < adjMatrix[u][v][1]) then
+        adjMatrix[u][v] := [w, idx];
+      fi;
+    od;
+  od;
+
+  # Store shortest paths for single edges
+  distances := EmptyPlist(nrVertices);
+  parents   := EmptyPlist(nrVertices);
+  edges     := EmptyPlist(nrVertices);
+  for u in vertices do
+    distances[u] := EmptyPlist(nrVertices);
+    parents[u]   := EmptyPlist(nrVertices);
+    edges[u]     := EmptyPlist(nrVertices);
+
+    for v in vertices do
+      distances[u][v] := infinity;
+      parents[u][v] := fail;
+      edges[u][v] := fail;
+
+      if u = v then
+        distances[u][v] := 0;
+        # if the same node, then the node has no parents
+        parents[u][v] := fail;
+        edges[u][v]   := fail;
+      elif IsBound(adjMatrix[u][v]) then
+        w   := adjMatrix[u][v][1];
+        idx := adjMatrix[u][v][2];
+
+        distances[u][v] := w;
+        parents[u][v]   := u;
+        edges[u][v]     := idx;
+      fi;
+    od;
+  od;
+
+  # try every triple: distance from u to v via k
```

Comment:
Can you please add a comment about why we aren't using the C Floyd-Warshall here? Just to understand again in the future, doesn't have to be very elaborate comment.